src/main/java/org/zalando/nakadi/service/kpi/AbstractNakadiPublisher.java

```diff
@@ -0,0 +1,28 @@
+package org.zalando.nakadi.service.kpi;
+
+import org.json.JSONObject;
+import org.zalando.nakadi.util.FeatureToggleService;
+
+public abstract class AbstractNakadiPublisher {
+
+    private final FeatureToggleService featureToggleService;
+    private final NakadiPublisher publisher;
+
+    protected AbstractNakadiPublisher(final FeatureToggleService featureToggleService,
+                                      final NakadiPublisher publisher) {
+        this.featureToggleService = featureToggleService;
+        this.publisher = publisher;
+    }
+
+    protected void publish(final String etName, final Object... data) {
+        if (!featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.KPI_COLLECTION)) {
+            return;
+        }
+
+        final JSONObject event = prepareData(data);
+        publisher.enrichAndSubmit(event, etName);
+    }
+
+    protected abstract JSONObject prepareData(final Object[] data);
```

Comment:
I don't really get why you decided to extract this to a separate method. Because of this you have to operate with this array of objects. I think the architecture will be more clear if the publish method would accept already JSONObject
`protected void publish(final String etName, final JSONObject data)`
And the child classes would have a separate publish method (as it is now) in which they will convert parameters to JSONObject and pass to this `publish`